File: rs/src/rpn_calculator.rs
use embedded_hal::serial::*;
use nb::block;

#[derive(Clone, Copy)]
enum Operator {
    Plus,
    Minus,
    Multiply,
    Divide,
    Remainder,
}

#[derive(Clone, Copy)]
enum Term {
    Value(isize),
    Operator(Operator),
}

enum Error {
    InvalidTerm,
    TooManyTerms,
    ReadError,
    TokenTooLong,
    StackOverflow,
    StackUnderflow,
}

struct Stack {
    values: [isize; STACK_SIZE],
    top_index: usize,
}

impl Default for Stack {
    fn default() -> Self {
        Stack {
            values: [0; STACK_SIZE],
            top_index: 0,
        }
    }
}

impl Stack {
    fn push(&mut self, value: isize) -> Result<(), Error> {
        if self.top_index == STACK_SIZE {
            Err(Error::StackOverflow)
        } else {
            self.values[self.top_index] = value;
            self.top_index += 1;
            Ok(())
        }
    }

    fn pop(&mut self) -> Result<isize, Error> {
        if self.top_index == 0 {
            Err(Error::StackUnderflow)
        } else {
            self.top_index -= 1;
            Ok(self.values[self.top_index])
        }
    }
}

const MAX_TERMS: usize = 32;
const MAX_TOKEN_LENGTH: usize = 16;
const STACK_SIZE: usize = 32;
const MAX_DIGITS: usize = 20;

const PROMPT: &'static str = "> ";

const SPACE: u8 = b" "[0];
const TAB: u8 = b"\t"[0];
const CARRIAGE_RETURN: u8 = b"\r"[0];
const LINE_FEED: u8 = b"\n"[0];
const ZERO: u8 = b"0"[0];
const NINE: u8 = b"9"[0];

pub fn run_repl<T: Read<u8> + Write<u8>>(port: &mut T) {
    let mut terms: [Term; MAX_TERMS] = [Term::Value(0); MAX_TERMS];
    loop {
        print(PROMPT, port);
        match read_expression(port, &mut terms).and_then(|n| evaluate(&terms, n)) {
            Ok(result) => print_result(result, port),
            Err(error) => print_error(error, port),
        }
    }
}

fn read_expression<T: Read<u8> + Write<u8>>(
    port: &mut T,
    terms: &mut [Term; MAX_TERMS],
) -> Result<usize, Error> {
    let mut token: [u8; MAX_TOKEN_LENGTH] = [0; MAX_TOKEN_LENGTH];
    let mut number_of_terms = 0;
    loop {
        let (token_length, end_of_line) = read_token(port, &mut token)?;
        let term = read_operator(&token, token_length)
            .or(read_value(&token, token_length))
            .ok_or(Error::InvalidTerm)?;

        if number_of_terms == MAX_TERMS {
            return Err(Error::TooManyTerms);
        } else {
            terms[number_of_terms] = term;
            number_of_terms += 1;
        }

        if end_of_line {
            break;
        }
    }
    Ok(number_of_terms)
}

fn evaluate(terms: &[Term; MAX_TERMS], number_of_terms: usize) -> Result<isize, Error> {
    let mut stack: Stack = Default::default();
    for index in 0..number_of_terms {
        match terms[index] {
            Term::Value(value) => stack.push(value)?,
            Term::Operator(operator) => {
                let second = stack.pop()?;
                let first = stack.pop()?;
                let result = apply(operator, first, second);
                stack.push(result)?;
            }
        };
    }
    stack.pop()
}

fn print_result(result: isize, output: &mut impl Write<u8>) {
    let mut remaining: usize = if result < 0 {
        print("-", output);
        (result * -1) as usize
    } else {
        result as usize
    };

    let mut digits: [u8; MAX_DIGITS] = [0; MAX_DIGITS];
    let mut number_of_digits = 0;
    loop {
        let digit_value = remaining % 10;
        digits[number_of_digits] = ZERO + digit_value as u8;
        remaining /= 10;
        number_of_digits += 1;

        if remaining == 0 {
            break;
        }
    }

    for index in (0..number_of_digits).rev() {
        print_byte(digits[index], output);
    }
    end_line(output);
}

fn print_error(error: Error, output: &mut impl Write<u8>) {
    let message = match error {
        Error::InvalidTerm => "a term in the expression was invalid",
        Error::TooManyTerms => "too many terms in the expression",
        Error::ReadError => "error while reading input",
        Error::TokenTooLong => "a term in the expression was too long",
        Error::StackOverflow => "the stack overflowed",
        Error::StackUnderflow => "the stack underflowed",
    };

    end_line(output);
    print("Error: ", output);
    print(message, output);
    end_line(output);
}

fn read_token<T: Read<u8> + Write<u8>>(
    port: &mut T,
    token: &mut [u8; MAX_TOKEN_LENGTH],
) -> Result<(usize, bool), Error> {
    let mut token_length = 0;
    loop {
        let byte = echoing_read(port)?;

        if byte == SPACE || byte == TAB {
            return Ok((token_length, false));
        } else if byte == CARRIAGE_RETURN {
            print_byte(LINE_FEED, port);
            return Ok((token_length, true));
        } else if token_length == MAX_TOKEN_LENGTH {
            return Err(Error::TokenTooLong);
        } else {
            token[token_length] = byte;
            token_length += 1;
        }
    }
}

fn read_operator(token: &[u8; MAX_TOKEN_LENGTH], token_length: usize) -> Option<Term> {
    if token_length > 1 {
        None
    } else {
        match token[0] as char {
            '+' => Some(Operator::Plus),
            '-' => Some(Operator::Minus),
            '*' => Some(Operator::Multiply),
            '/' => Some(Operator::Divide),
            '%' => Some(Operator::Remainder),
            _ => None,
        }
        .map(|operator| Term::Operator(operator))
    }
}

fn read_value(token: &[u8; MAX_TOKEN_LENGTH], token_length: usize) -> Option<Term> {
    let (sign, digits_start) = if token[0] as char == '-' {
        (-1, 1)
    } else {
        (1, 0)
    };

    let mut magnitude: isize = 0;
    for index in digits_start..token_length {
        if token[index] < ZERO || token[index] > NINE {
            return None;
        }
        magnitude *= 10;
        magnitude += (token[index] - ZERO) as isize;
    }

    Some(Term::Value(sign * magnitude))
}

fn apply(operator: Operator, first: isize, second: isize) -> isize {
    match operator {
        Operator::Plus => first + second,
        Operator::Minus => first - second,
        Operator::Multiply => first * second,
        Operator::Divide => first / second,
        Operator::Remainder => first % second,
    }
}

fn print(message: &str, output: &mut impl Write<u8>) {
    for &byte in message.as_bytes() {
        print_byte(byte, output);
    }
}

fn print_byte(byte: u8, output: &mut impl Write<u8>) {
    block!(output.write(byte)).unwrap_or_default();
}

fn end_line(output: &mut impl Write<u8>) {
    print_byte(CARRIAGE_RETURN, output);
    print_byte(LINE_FEED, output);
}

fn echoing_read<T: Read<u8> + Write<u8>>(port: &mut T) -> Result<u8, Error> {
    let byte = block!(port.read()).map_err(|_| Error::ReadError)?;
    print_byte(byte, port);
    Ok(byte)
}

File: c/src/rpn_calculator.c
#include "rpn_calculator.h"

#include "io.h"

#define PROMPT "> "
#define MAX_TERMS 32
#define MAX_DIGITS 20
#define MAX_TOKEN_LENGTH 16
#define STACK_SIZE 32

enum error {
	NO_ERROR,
	INVALID_TERM,
	TOO_MANY_TERMS,
	TOKEN_TOO_LONG,
	STACK_OVERFLOW,
	STACK_UNDERFLOW,
};

enum term_type {
	VALUE,
	OPERATOR,
};

enum operator{
	PLUS,
	MINUS,
	MULTIPLY,
	DIVIDE,
	REMAINDER,
};

struct term {
	enum term_type type;
	union {
		int value;
		enum operator operator;
	};
};

struct expression {
	struct term terms[MAX_TERMS];
	unsigned term_count;
};

struct stack {
	int values[STACK_SIZE];
	unsigned top_index;
};

static const char *crlf = "\r\n";
static const char *error_messages[] = {
	[INVALID_TERM] = "a term in the expression was invalid",
	[TOO_MANY_TERMS] = "too many terms in the expression",
	[TOKEN_TOO_LONG] = "a term in the expression was too long",
	[STACK_OVERFLOW] = "the stack overflowed",
	[STACK_UNDERFLOW] = "the stack underflowed",
};

static enum error read_expression(struct expression *expression_out);
static enum error evaluate(
	const struct expression *expression, int *result_out);

static void print_error(enum error e);
static void print_result(int result);
static enum error read_token(char token[MAX_TOKEN_LENGTH],
	unsigned *token_length_out, short *end_of_line_out);
static short try_read_value(
	char token[MAX_TOKEN_LENGTH], unsigned token_length, struct term *term_out);
static short try_read_operator(
	char token[MAX_TOKEN_LENGTH], unsigned token_length, struct term *term_out);
static enum error push(struct stack *stack, int value);
static enum error pop(struct stack *stack, int *value_out);
static int apply(enum operator operator, int first, int second);

void run_repl(void)
{
	static struct expression expression;
	enum error e;
	int result;

	while (1) {
		print_str(PROMPT);
		if ((e = read_expression(&expression)) != NO_ERROR
			|| (e = evaluate(&expression, &result)) != NO_ERROR)
			print_error(e);
		else
			print_result(result);
	}
}

static enum error read_expression(struct expression *expression_out)
{
	expression_out->term_count = 0;

	static char token[MAX_TOKEN_LENGTH];
	unsigned token_length;
	short end_of_line;
	enum error e;

	while (1) {
		if ((e = read_token(token, &token_length, &end_of_line)) != NO_ERROR)
			return e;

		if (token_length == 0)
			continue;

		struct term *current_term
			= &expression_out->terms[expression_out->term_count];
		if (try_read_operator(token, token_length, current_term)
			|| try_read_value(token, token_length, current_term))
			++expression_out->term_count;
		else
			return INVALID_TERM;

		if (expression_out->term_count == MAX_TERMS)
			return TOO_MANY_TERMS;

		if (end_of_line)
			break;
	}

	return NO_ERROR;
}

static enum error evaluate(const struct expression *expression, int *result_out)
{
	int first, second;
	struct stack stack = { 0 };
	enum error e;
	for (unsigned i = 0; i < expression->term_count; ++i) {
		switch (expression->terms[i].type) {
		case VALUE:
			if ((e = push(&stack, expression->terms[i].value)) != NO_ERROR)
				return e;
			break;

		case OPERATOR:
			if ((e = pop(&stack, &second)) != NO_ERROR)
				return e;
			if ((e = pop(&stack, &first)) != NO_ERROR)
				return e;
			push(&stack, apply(expression->terms[i].operator, first, second));
			break;
		}
	}

	return pop(&stack, result_out);
}

static void print_error(enum error e)
{
	print_str(crlf);
	print_str("Error: ");
	print_str(error_messages[e]);
	print_str(crlf);
}

static void print_result(int result)
{
	if (result < 0) {
		print_char('-');
		result *= -1;
	}

	char digits[MAX_DIGITS];
	unsigned n = 0;
	do {
		digits[n++] = '0' + result % 10;
		result /= 10;
	} while (result != 0);

	while (--n < MAX_DIGITS)
		print_char(digits[n]);

	print_str(crlf);
}

static enum error read_token(char token[MAX_TOKEN_LENGTH],
	unsigned *token_length_out, short *end_of_line_out)
{
	char c;
	unsigned length = 0;
	while (1) {
		switch (c = read_char()) {
		case ' ':
		case '\t':
			*token_length_out = length;
			*end_of_line_out = 0;
			return NO_ERROR;

		case '\r':
			print_char('\n');
			*token_length_out = length;
			*end_of_line_out = 1;
			return NO_ERROR;

		default:
			if (length == MAX_TOKEN_LENGTH)
				return TOKEN_TOO_LONG;
			else
				token[length++] = c;
		}
	}
	return NO_ERROR;
}

static short try_read_value(
	char token[MAX_TOKEN_LENGTH], unsigned token_length, struct term *term_out)
{
	int sign;
	unsigned digits_start;
	if (token[0] == '-') {
		sign = -1;
		digits_start = 1;
	} else {
		sign = 1;
		digits_start = 0;
	}

	int magnitude = 0;
	for (unsigned i = digits_start; i < token_length; ++i) {
		if (token[i] < '0' || token[i] > '9')
			return 0;
		magnitude *= 10;
		magnitude += token[i] - '0';
	}

	term_out->type = VALUE;
	term_out->value = sign * magnitude;
	return 1;
}

static short try_read_operator(
	char token[MAX_TOKEN_LENGTH], unsigned token_length, struct term *term_out)
{
	if (token_length != 1)
		return 0;

	enum operator operator;
	switch (token[0]) {
	case '+':
		operator= PLUS;
		break;
	case '-':
		operator= MINUS;
		break;
	case '*':
		operator= MULTIPLY;
		break;
	case '/':
		operator= DIVIDE;
		break;
	case '%':
		operator= REMAINDER;
		break;

	default:
		return 0;
	}

	term_out->type = OPERATOR;
	term_out->operator= operator;
	return 1;
}

static enum error push(struct stack *stack, int value)
{
	if (stack->top_index == STACK_SIZE)
		return STACK_OVERFLOW;
	stack->values[stack->top_index++] = value;
	return NO_ERROR;
}

static enum error pop(struct stack *stack, int *value_out)
{
	if (stack->top_index == 0)
		return STACK_UNDERFLOW;
	*value_out = stack->values[--stack->top_index];
	return NO_ERROR;
}

static int apply(enum operator operator, int first, int second)
{
	switch (operator) {
	case PLUS:
		return first + second;
	case MINUS:
		return first - second;
	case MULTIPLY:
		return first * second;
	case DIVIDE:
		return first / second;
	case REMAINDER:
		return first % second;

	default:
		return 0;
	}
}
